### PR TITLE
Added confusion matrix op, issue #3637

### PR DESCRIPTION
**Pull request by RadhikaG**
Hi, I've added the confusion matrix operation from https://github.com/Theano/Theano/issues/3637 to `theano/tensor/nnet/nnet.py`, since it's frequently used in Machine Learning and I assumed this to be the correct file. Please let me know if I need to change the location of the operation, and I'll accordingly add tests for the op to the appropriate file.

Thanks!

### CODE REVIEW

**Comment by nouiz** [comment]
Thanks for the PR. I think we should not make a new OP for this, but build a Theano graph of existing ops. This would allow them to be optimized by Theano (like using c code) and to work on GPU. The GPU isn't for the speed, but to remove the transfer to/from GPU.

I think you can do so with mostly similar code then what is in the perform of this op. Do you want to try that?

**Comment by RadhikaG** [comment]
No problem. I'll implement just a function for the confusion matrix using existing Theano ops. I assume the location of the confusion matrix function is alright?

**Comment by RadhikaG** [comment]
Also, I noticed that there is no union1d operation for tensors. Shall I add on the union1d op in a separate PR as a continuation of https://github.com/Theano/Theano/issues/1916?

**Comment by RadhikaG** [comment]
I've refactored the confusion matrix op to a function, and added unit tests. Please let me know if I need to make any more changes.

**Comment by RadhikaG** [comment]
One Travis [job](https://travis-ci.org/Theano/Theano/jobs/111837509) is still failing, even though it passes on my local install. Could you please retest it once?

**Comment by nouiz** [comment]
this travis problem happen randomly. I restarted it.

**Comment by trungnt13** [comment]
Is it better to support also 2-d tensor as input (softmax matrix from the network)? 
For example:
```
    if pred.ndim == 2:
        pred = tensor.argmax(pred, axis=-1)
    elif pred.ndim != 1:
        raise ValueError('pred must be 1-d or 2-d tensor variable')
```

**Comment by nouiz** [comment]
I think the argmax should be outside this function. But as always, we can
discuss this.

Why do you think it should be inside?

**Comment by nouiz** [comment]
I like the idea of having the same interface as scikit.learn. But there isn't a big difference.

I also think the problem of small batch is important. As the classes much be integer starting from 0 with the current code, what about passing the parameter n_classes that is the number of classes. I think it could be a mandatory parameter. What do you think?

What scikit.learn do when classes isn't passed?

**Comment by trungnt13** [comment]
This is the interface from sklearn:

```
y_true = ["cat", "ant", "cat", "cat", "ant", "bird"]
y_pred = ["ant", "ant", "cat", "cat", "ant", "cat"]
confusion_matrix(y_true, y_pred, labels=["ant", "bird", "cat"])
# or
y_true = [2, 0, 2, 2, 0, 1]
y_pred = [0, 0, 2, 2, 0, 2]
confusion_matrix(y_true, y_pred) # labels=[0,1,2] in this case
```

However, I think in theano case, labels should be a number (nb_classes) or a list of number represent class index (in case users can constraint the order of row and column of confusion matrix). 

Frankly speaking, I prefer a mandatory arguments _nb_classes_ as this is my only use case.

For the _argmax_ idea, just for convenient, I cannot think about the case users accidentally give 2D input to _confusion_matrix_ that they expect the methods do something else instead of argmax.

**Comment by nouiz** [comment]
As Theano don't handle string, I think we must request numbers for inputs.
I would suggest that we requeset inputs between 0 and n classe -1. With
this interface, the 3rd parameter can be n_classes.

What do you think of that? It is mostly the conversion to number of that
interface. We could use another definition of classe number, like start at
0, but I always  see peple start from 0 in deep learning.

**Comment by trungnt13** [comment]
I agree with your interface, so the final interface should be:
> confusion_matrix(pred, actual, nb_class) # tensor.extra_ops.to_one_hot also have nb_class argument

I prefer restrict all inputs must start from 0 to (nb_class -1). 

The most important thing is we have the funciton in theano as soon as possible.

**Comment by nouiz** [comment]
So we agree on the interface, that is great:)
This branch also need a rebase before being merged.

thanks.

**Comment by nouiz** [comment]
Does someone have time to finish this PR? It would be great to get this merged.

**Comment by xoltar** [comment]
@nouiz I rebased and updated docstring example, new PR is  #4810 .

**Comment by nouiz** [comment]
closing as continued in gh-4810
